torch/_inductor/fx_passes/post_grad.py

```diff
@@ -484,6 +576,125 @@ def is_valid_splitwithsizes_cat(match):
+
+    input_storages = set()
+    output_storages = set()
+    for node in graph.nodes:
+        if node.op == "placeholder":
+            input_storages.add(get_node_storage(node))
+        else:
+            break
+
+    for out in tuple(graph.nodes)[-1].args[0]:
+        if isinstance(out, torch.fx.Node):
+            output_storages.add(get_node_storage(out))
+
+    noop_conditions = {
+        aten.slice.Tensor: (0, slice_cond),
+        aten.clone.default: (0, true),
+        aten.slice_scatter.default: (1, slice_scatter_cond),
+    }
+
+    for node in graph.nodes:
+        if node.target in noop_conditions:
+            cloned_arg_index, cond = noop_conditions[node.target]
+            src = node.args[cloned_arg_index]
+            if (
+                get_node_storage(node) in output_storages
+                and get_node_storage(src) in input_storages
+            ):
+                continue
+            if same_layout(node, src) and cond(*node.args, **node.kwargs):
+                node.replace_all_uses_with(src)
+                graph.erase_node(node)
+
+
+def reinplace_scatters(graph):
+    """
+    Reinplaces scatter operations in easy cases where the node being mutated
+    is only used by the scatter (users == 1), and the node being mutated
+    shares storage with no other nodes.
+
+    Also handles input mutations when there is a corresponding copy node.
+    """
+
+    copy_nodes = {}
+    mutated_inputs = set()
+    storage_to_nodes = defaultdict(list)
+    for node in reversed(graph.nodes):
+        if isinstance(node.meta["val"], torch.Tensor):
+            storage_to_nodes[get_node_storage(node)].append(node)
+        if node.target == aten.copy_.default:
+            copy_nodes[(node.args[0], node.args[1])] = node
+            mutated_inputs.add(node.args[0])
+        elif node.op == "output":
+            pass
+
+    InplaceableOp = namedtuple("InplaceableOp", ["inplace_op", "mutated_arg"])
+
+    inplaceable_ops = {
+        aten.index_put.default: InplaceableOp(aten.index_put_.default, 0),
+    }
+    for node in graph.nodes:
+        if inplaceable_op := inplaceable_ops.get(node.target, False):
+            mutated_arg = node.args[inplaceable_op.mutated_arg]
+            if get_node_storage(mutated_arg) is None:
+                continue
+            shared_view_nodes = storage_to_nodes[get_node_storage(mutated_arg)]
+            if mutated_arg.op == "placeholder":
+                if not (copy_node := copy_nodes.get((mutated_arg, node), False)):
+                    continue
+
+                if (
+                    len(shared_view_nodes) > 2
+                ):  # Arg aliases another node other than copy_
+                    continue
+
+                # Check for any uses other than current node and copy_ epilogue
+                if len(mutated_arg.users) > 2:
+                    continue
+
+                graph.erase_node(copy_node)
+                node.target = inplaceable_op.inplace_op
```

Comment:
```
def foo(y):
     x = index_put(y, ...)
     z = x + 1
     y.copy_(x)
     return y, z
```

Do we have any checks that `len(node.users) == 1` ? We have checks on both args but not on the output.

Reply:
This is a valid optimization to make in this case, no?